include/RAJA/pattern/params/reducer.hpp

```diff
@@ -83,16 +55,51 @@ namespace detail
   // Basic Reducer
   //
   //
-  template <typename Op, typename T>
+
+  // For the standard use case, T can be a basic data type or a ValLoc, and DataType is the same as T.
+  // For the pass through use case, T is a ValLoc, and DataType is the underlying basic data type within the ValLoc.
+  // By default, we expect the standard use case, which is why passthru is set to false.
+  template <template <typename, typename, typename> class Op, typename T, typename VType, typename DataType = T, typename IndexType = RAJA::Index_type, bool passthru = false>
   struct Reducer : public ForallParamBase {
```

Comment:
Could this be partially specialized for ValLoc instead of having both in one definition?

Reply:
Maybe I can document this better in the comments, but there are actually 3 cases here:

1. T = basic data type declared by the user, e.g. `Reduce<>(myint)`.
2. T = ValLoc declared by the user, e.g. `ReduceLoc<>(myvalloc)`.
3. T = ValLoc declared by RAJA, and DataType = basic data type declared by the user (and IndexType declared by the user), e.g. `ReduceLoc<>(data, index)`.

1 and 2 are both covered with the prior code, and probably should not be separated. 3 adds the complexity of the pass through variables, and would benefit from some form of specialization but it's not really a partial specialization case. Perhaps I can revert `Reducer` to its old code that handles 1 and 2, and add a `ReducerLoc` struct which inherits `Reducer`; does that sound reasonable?

Reply:
@MrBurmark This has been updated with your partial specialization suggestion.